### PR TITLE
(Closes #313) permit intrinsic shadowing

### LINKED ISSUE #313: Allow for overriding of intrinsics with different numbers of arguments
If code shadows an intrinsic but gives it a different interface then fparser currently raises a SyntaxError as the interface does not match that of the Fortran intrinsic and it does not countenance any other options.
It would probably be better if the match on the intrinsic were simply allowed to fail at this stage.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
This has been needed for NEMOVAR for a while: currently we are over zealous in checking the number of arguments to intrinsics and fell foul of the fact that NEMOVAR overrides DOT_PRODUCT with something that can accept more args than the Fortran intrinsic can. This PR alters the parser's behaviour so that if the number of arguments don't match and there's somewhere that a symbol could be coming from then we don't match an intrinsic. I had to extend SymbolTable with a new `wildcard_imports` property to do this.
One for @sergisiso or @rupertford.

File: src/fparser/two/tests/fortran2003/test_intrinsics.py
Comment (rupertford) [inline src/fparser/two/tests/fortran2003/test_intrinsics.py:297]: You could add some other cases here where the use statement is somewhere else e.g. in the subroutine and any other cases you can think of.

File: src/fparser/two/tests/test_symbol_table.py
Comment (rupertford) [inline src/fparser/two/tests/test_symbol_table.py:282]: Please update the date in the file.

File: src/fparser/two/tests/test_symbol_table.py
Comment (rupertford) [inline src/fparser/two/tests/test_symbol_table.py:288]: Do you have a test that checks that the logic works if the shadowed intrinsic is named in a use statement. I can't remember. I'm sure you already do.

File: src/fparser/two/tests/test_symbol_table.py
Comment (rupertford) [inline src/fparser/two/tests/test_symbol_table.py:283]: Good stuff, but does it work in other situations.

**Review by rupertford — CHANGES_REQUESTED** [review]
Good job. I have mostly minor comments but you will see that I am a little concerned as to whether this solution works in situations other than module/subroutine.

**Comment by arporter** [comment]
I hadn't considered submodules so thanks for spotting that. Having done some reading I see that (as you'd expect) they have access to all the symbols made available by the 'parent' module. I'll have to allow for that.

File: src/fparser/two/symbol_table.py
Comment (arporter) [reply src/fparser/two/symbol_table.py:704]: Done.

File: src/fparser/two/symbol_table.py
Comment (arporter) [reply src/fparser/two/symbol_table.py:715]: I don't believe there's any limit to the hierarchy of scope in Fortran. submodules however are a different matter and I'll have to handle those somehow.

Comment (arporter) [reply src/fparser/two/symbol_table.py:715]: Comment added.

File: src/fparser/two/tests/fortran2003/test_intrinsics.py
Comment (arporter) [reply src/fparser/two/tests/fortran2003/test_intrinsics.py:298]: Oops.

File: src/fparser/two/tests/fortran2003/test_intrinsics.py
Comment (arporter) [reply src/fparser/two/tests/fortran2003/test_intrinsics.py:297]: I've made it a bit more complete. I'll add a test for a submodule as well.

**Comment by arporter** [comment]
I've hit a problem with submodules which I think will be common to any situation where we need to do a context-sensitive match. At the moment, any `match` method knows only the string it is trying to match - it doesn't know what node (if any) will be the parent of the one it is creating. This is a problem here because I was going to allow intrinsic shadowing in a submodule on the basis that we assume the Fortran is valid and therefore the necessary symbol must be being brought into scope in the parent module (which we can't see because it is in a separate file).

**Comment by arporter** [comment]
I can solve my immediate problem by extending the SymbolTable so that (as in the PSyIR) it keeps a reference to the node in the parse tree with which it is associated. This is probably not a bad thing to do and is considerably lighter weight than changing the way we construct nodes. I'll do it as a separate PR though so we can have a proper discussion.

**Comment by rupertford** [comment]
Would your symbol table implementation help here? I'm not sure how hard it would be to create a list of nodes during a hierarchical match. Give the current implementation I would expect it to be difficult - assuming that is what you are after.

**Comment by arporter** [comment]
It will definitely let me determine whether we are currently within a `submodule` - I had a little play and it was easy.

**Comment by rupertford** [comment]
Great. I think the problem with ancestor is that we create the nodes as we recurse back up, not down - but perhaps there is a simple solution.

**Comment by arporter** [comment]
> Great. I think the problem with ancestor is that we create the nodes as we recurse back up, not down - but perhaps there is a simple solution.

Ah! You are of course correct. I only checked for the immediate parent (which works) but the grandparent isn't created yet :-(

**Comment by arporter** [comment]
Perhaps we revert to the assumption that the Fortran is correct so that if the number of arguments is invalid then we simply don't match an Intrinsic?

**Comment by rupertford** [comment]
When would this allow incorrect code to get through?

**Comment by arporter** [comment]
> When would this allow incorrect code to get through?

If the Fortran code *was* incorrect (so meant to call an intrinsic but supplied the wrong number of args) then we wouldn't spot it - it would get matched as a call or array access instead.

Since we don't check argument types, we will also incorrectly match an intrinsic call when a Fortran compiler would complain if the wrong type was passed to an intrinsic.

**Comment by arporter** [comment]
If we are parsing the following code:

      submodule (foobar) bar

        contains
        subroutine my_sub
          real :: a
          a = dot_product(1,2,3)
        end subroutine my_sub
      end

then at the point when we are trying to match `dot_product`, the SymbolTable approach means that we know we are within a subroutine but that's all we know. To know more than that we probably need some new data structure that keeps track of our current position in the matching hierarchy (i.e. the chain of classes we are currently trying to match).
That's probably not that hard to do but feels like possibly a retrograde step c.f. #191.

**Comment by rupertford** [comment]
Hi @arporter, some thoughts.

Could we not have some more context from the symbol table by capturing more within it, i.e. know that the subroutine is within a submodule, or do we not know that yet due to the recursion? I'm not saying it would solve the problem, I was just wondering.

I agree the proposal in in #191 sounds great, it is just a question of effort (and trust in the new implementation I guess). Do you think that it would make sense for us to take this on, given there is no funding associated with this?

We could optionally require module information (in some standard form) from other source code if it could not be found in addition to the current implementation I suppose, but I don't think it should be mandatory otherwise we can no longer parse any code in isolation, which is the main use of the parser.

**Comment by arporter** [comment]
Good news! I must have got confused when I said that storing the scoping node in the SymbolTable didn't help because I was wrong. I've also added a method to the SymbolTable that reports whether it's possible for a Symbol to be brought into scope from elsewhere. With those changes I have the tests working, including the one with a submodule.

File: src/fparser/two/tests/test_symbol_table.py
Comment (arporter) [reply src/fparser/two/tests/test_symbol_table.py:282]: Done.

File: src/fparser/two/tests/test_symbol_table.py
Comment (arporter) [reply src/fparser/two/tests/test_symbol_table.py:288]: I was about to confidently reply "yes" but it turns out that I didn't. I can't quite believe it. Anyway, that then showed that SymbolTable.lookup() wasn't checking for imported Symbols so I've updated it.

File: src/fparser/two/tests/test_symbol_table.py
Comment (arporter) [reply src/fparser/two/tests/test_symbol_table.py:301]: Extended.

**Comment by arporter** [comment]
This PR has got somewhat larger: the SymbolTable now keeps a reference to the parse tree node with which it is associated. I've also extended the 'lookup' functionality so that it includes symbols imported from other modules. I've also added a "all_symbols_resolved" query method that establishes whether any symbols could be being brought into scope from elsewhere.
Ready for another look.

**Comment by arporter** [comment]
@rupertford happily (for me) it turns out this is in fact waiting on a review from yourself :-)

File: src/fparser/two/symbol_table.py
Comment (rupertford) [reply src/fparser/two/symbol_table.py:757]: This seems to have been superseded by other changes

File: src/fparser/two/tests/fortran2003/test_intrinsics.py
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_intrinsics.py:298]: This seems to have been superseded by other changes.

File: src/fparser/two/tests/fortran2003/test_intrinsics.py
Comment (rupertford) [reply src/fparser/two/tests/fortran2003/test_intrinsics.py]: There was no submodule example but I expect that is either explained or dealt with later.

File: src/fparser/two/tests/test_symbol_table.py
Comment (rupertford) [reply src/fparser/two/tests/test_symbol_table.py:306]: I agree that you do now :-)

File: src/fparser/two/tests/test_symbol_table.py
Comment (rupertford) [reply src/fparser/two/tests/test_symbol_table.py:301]: Thanks.

File: src/fparser/two/symbol_table.py
Comment (rupertford) [inline src/fparser/two/symbol_table.py:640]: It's not clear to me why you test parent scope and only parent scope in the code below. Could you please explain in a comment. Is it to do with routines being inside modules?

File: src/fparser/two/symbol_table.py
Comment (rupertford) [inline src/fparser/two/symbol_table.py:772]: If a variable is not used in a routine, do we actually care if it is resolved or not? I'm not saying we should change anything as I understand this is from the perspective of the symbol_table but we could potentially ignore/remove symbols that are declared but not used, or wildcard imports if all symbols that are used are already resolved.

File: src/fparser/two/tests/test_module_use.py
Comment (rupertford) [inline src/fparser/two/tests/test_module_use.py:163]: Please update the date.

**Review by rupertford — CHANGES_REQUESTED** [review]
Looks good. Just some minor requests.

File: src/fparser/two/symbol_table.py
Comment (arporter) [reply src/fparser/two/symbol_table.py:166]: I've just tried removing it and Black put it back so no, but it needs to be there :-)

File: src/fparser/two/symbol_table.py
Comment (arporter) [reply src/fparser/two/symbol_table.py:433]: I've gone for the latter.

File: src/fparser/two/symbol_table.py
Comment (arporter) [reply src/fparser/two/symbol_table.py:640]: Oh yes, that whole method was rather comment-light. Essentially we recurse upwards (`self.parent.lookup` calls this method on the parent symbol table) - I've added a comment.

File: src/fparser/two/symbol_table.py
Comment (arporter) [reply src/fparser/two/symbol_table.py:772]: We could potentially remove symbols that aren't used although, as you say, this method is about whether or not there could be symbols being brought into scope from elsewhere. I tried to think of a better name for it but failed. However, I think the docstring is pretty clear.

**Comment by arporter** [comment]
Ready for another look now...

**Review by rupertford — CHANGES_REQUESTED** [review]
Changes look good. However, there appears to be a line that is not covered by the tests. Could you please check and fix if required.

**Comment by arporter** [comment]
Thanks for spotting that missed line - I'd been fooled by Codecov reporting success. Happily that line was associated with #170 (which is now fixed) and so I've been able to simplify it.
Ready for another look.

**Review by sergisiso — APPROVED** [review]
As discussed with @arporter I took over this review in order to add this functionality to the next release. I see that @rupertford was already happy with how his previous comments were addressed and just one line on the codecov was missing. So I just gave it a quick read  and checked that the codecov is now complete. This PR is ready to merge.

**Pull request merged**
